docroot/themes/humsci/humsci_basic/src/js/shared/media/media-caption-toggle.js

```diff
@@ -0,0 +1,103 @@
+
+          // Condition to add toggle
+          const addToggle = actualHeight >= 38 || (parentSpotlight && viewportWidth <= 991);
+
+          // Condition to remove toggle
+          const removeToggle = !addToggle;
+
+          // Add toggle structure
+          if (addToggle && !existingWrapper) {
+            caption.classList.add('toggle-caption__wrapper');
+
+            // Create wrapper
+            const textWrapper = document.createElement('div');
+            textWrapper.className = 'toggle-caption__content';
+            textWrapper.style.visibility = 'hidden';
+            caption.appendChild(textWrapper);
+
+            // Create close button (inside wrapper, before text)
+            const closeBtn = document.createElement('button');
+            closeBtn.className = 'toggle-caption__close';
+            closeBtn.setAttribute('aria-hidden', 'true');
+            closeBtn.style.visibility = 'hidden';
+            textWrapper.appendChild(closeBtn);
+
+            // Move text inside wrapper
+            captionText.forEach((el) => textWrapper.appendChild(el));
+
+            // Create toggle button (outside wrapper)
+            const toggle = document.createElement('button');
+            toggle.className = 'toggle-caption__toggle';
+            toggle.setAttribute('aria-hidden', 'true');
+            caption.insertBefore(toggle, textWrapper);
+
+            // Toggle behavior
+            toggle.addEventListener('click', () => {
+              toggle.style.display = 'none';
+              textWrapper.style.visibility = 'visible';
+              closeBtn.style.visibility = 'visible';
+            });
+
+            closeBtn.addEventListener('click', () => {
+              textWrapper.style.visibility = 'hidden';
+              closeBtn.style.visibility = 'hidden';
+              toggle.style.display = 'block';
+            });
+          }
+
+          // Remove toggle structure
+          if (removeToggle && existingWrapper) {
+            caption.classList.remove('toggle-caption__wrapper');
+
+            const toggleBtn = caption.querySelector('.toggle-caption__toggle');
+            const closeBtn = caption.querySelector('.toggle-caption__close');
+            const textWrapper = caption.querySelector('.toggle-caption__content');
+
+            // Move caption text elements back to main caption
+            const innerElements = Array.from(textWrapper.children).filter(
+              (el) => !el.classList.contains('toggle-caption__close'),
+            );
+            innerElements.forEach((el) => caption.appendChild(el));
+
+            // Remove created buttons and wrapper
+            if (toggleBtn) toggleBtn.remove();
+            if (closeBtn) closeBtn.remove();
+            if (textWrapper) textWrapper.remove();
+          }
+        });
+      };
+
+      // Initial run
+      processCaptions();
+
+      // Run again on viewport resize
+      let resizeTimeout;
+      window.addEventListener('resize', () => {
```

Comment:
Explore using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) for the individual captions instead of a global resize event.

Reply:
I tried to use it but there was something causing a loop on captions that were longer in large screensizes. I couldn't find what and why it was happening. And it was specifically with this function.

Reply:
I tried to implement it by changing the logic it was getting more complicated than expected. `ResizeObserver` observes the rendered size of an element (offsetHeight, contentRect, etc.). By adding or removing classes (like `visually-hidden` or `collapsible-caption`) it changes the layout of the element, and each layout change triggers the observer again.

Example:
- `updateCaptionState` adds `visually-hidden` because the caption is collapsible.
- The browser recalculates layout, so ResizeObserver fires again.
- `updateCaptionState` runs again, it add/remove classes again.
- This creates an infinite loop, especially because offsetHeight keeps changing slightly due to rounding or wrapping.